### DISCUSSION TITLE
Copy and paste in egui

### DISCUSSION THREAD

**Discussion by itsNiccoloSabatini** [emilk/egui#4516] [Q&A] [answered]
hi everyone, can someone help me?
I'm trying to copy and paste a text from my page using ctrl+c and ctrl+v and it doesn't do anything, i also tried the button method that someone else already discussed about. Can someone tell me how it works??

thanks

**Comment by itsNiccoloSabatini** [comment]
`if ui.button("📋").clicked() {
                    ui.output_mut(|o| {
                        warn!("Copying API key to clipboard: {:?}", key);
                        o.copied_text = key.clone();
                        warn!("Copied API key to clipboard: {:?}", o.copied_text);
                    });
                }`

i'm using this code to try copy, the `o.copied_text` is populated but there's nothing into the clipboard

**Comment by YgorSouza** [comment]
If you are building for WASM, you should add this to your project:

https://github.com/emilk/egui/blob/262a8bcf98870a382cc4a4569719366e1fcad585/.cargo/config.toml#L1-L6

**Reply by itsNiccoloSabatini** [reply]
i already have this in the same file and folder inside my project:
![Schermata del 2024-05-24 11-55-14](https://github.com/emilk/egui/assets/126770852/6f4a22f3-cb58-4518-977e-b3c7643bd3d2)

**Reply by itsNiccoloSabatini** [reply]
i want to give an update, using `ctx.inspection_ui(ui) ` for inspect what event are happening, so first of all : 
![Schermata del 2024-05-24 15-38-24](https://github.com/emilk/egui/assets/126770852/30d34b60-b856-46b3-9390-b44459c6bc42)
the text is recognized as a text to copy but when i try to copy nothing is been copied:
![immagine](https://github.com/emilk/egui/assets/126770852/c81f54c2-cc7e-4ea4-a9da-30fb7d3e9488)
and in  www.egui.rs is the same but it copy the text into the clipboard, in my project no:
![immagine](https://github.com/emilk/egui/assets/126770852/4fc4212f-2420-44e3-8714-04e9d8ca8038)

i used the "eframe_template" as a start point to develop this, any idea of what i'm missing?

**Reply by YgorSouza** [reply]
If you just `trunk serve` eframe_template directly without making any changes, are you able to copy the text from the page?

**Reply by itsNiccoloSabatini** [reply]
i'm going to try, any other idea for now?

**Reply by itsNiccoloSabatini** [reply]
So, really big update.
the reason why copy/paste wasn't working is that the site must be HTTPS to work and for my interface i need HTTP so that's the problem. i'm trying to find a solution.

**Reply by YgorSouza** [reply]
Really? Where did you see that? I mean it works as HTTP from localhost, doesn't it?

**Reply by emilk** [reply, marked as answer]
It may be that copy-and-paste only work in secure contexts:

* https://developer.mozilla.org/en-US/docs/Web/API/Window/isSecureContext
* https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts

**Reply by YgorSouza** [reply]
Yeah, that does seem to be the case.

> Secure context: This feature is available only in [secure contexts](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts) (HTTPS), in some or all [supporting browsers](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard#browser_compatibility).

**Reply by itsNiccoloSabatini** [reply]
> Yeah, that does seem to be the case.
> 
> > Secure context: This feature is available only in [secure contexts](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts) (HTTPS), in some or all [supporting browsers](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard#browser_compatibility).

yes is that. we tried the HTTPS and it worked as it should. so yes the solution is **USE HTTPS TO USE CLIPBOARD FEATURES**
